### ISSUE TITLE
[Default UI] BUG : Manually add Thing to Bridge => but Bridge uid NOT included in the Thing uid

### ISSUE THREAD

**Issue by andrewfg** [openhab/openhab-webui#290]
There is a bug in PaperUI **_EDIT: and in Default UI_**. When you try to add a Thing manually to an existing Bridge, the created Thing's UID does not include the UID of the respective bridge; and it's respective status is shown as `ONLINE`.

Note: we have tested this with the Hue, HD PowerView, Velux and NeoHub bindings, and the problem is the same; so it is NOT a binding problem.

 The steps to create the bug are shown in the series of screenshots below. 
---
![2020-06-25_19-50-02](https://user-images.githubusercontent.com/893994/85783260-0e4cba00-b71f-11ea-9f11-2c45bd048842.jpg)
---
![2020-06-25_19-51-09](https://user-images.githubusercontent.com/893994/85783289-160c5e80-b71f-11ea-960a-f894083af990.jpg)
---
![2020-06-25_19-55-44](https://user-images.githubusercontent.com/893994/85783388-2de3e280-b71f-11ea-9118-4fc48605edf0.jpg)
---
![2020-06-25_19-51-35](https://user-images.githubusercontent.com/893994/85783313-1c023f80-b71f-11ea-97b5-29f3d04698d1.jpg)
---
![2020-06-25_19-52-23](https://user-images.githubusercontent.com/893994/85783336-202e5d00-b71f-11ea-8b85-78c1f133454c.jpg)
---
![2020-06-25_19-53-03](https://user-images.githubusercontent.com/893994/85783362-27556b00-b71f-11ea-8213-7b3e03c34ee5.jpg)

**Comment by andrewfg** [comment]
Just for reference, below is a Wireshark capture. You can see that the Thing has been created without the bridge ID segment in UID; the UID should be `"UID": "hdpowerview:shade:**g24**:fdb5d366"`, but actually it generated `"UID": "hdpowerview:shade:fdb5d366"` with a missing `g24` third segment..


```
POST /rest/things HTTP/1.1
Host: 192.168.1.131:8080
Connection: keep-alive
Content-Length: 246
Accept: application/json, text/plain, */*
Accept-Language: en
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.116 Safari/537.36
Content-Type: application/json
Origin: http://192.168.1.131:8080
Referer: http://192.168.1.131:8080/paperui/index.html
Accept-Encoding: gzip, deflate

{
	"UID": "hdpowerview:shade:fdb5d366",
	"thingTypeUID": "hdpowerview:shade",
	"configuration": {
		"id": "yyyy"
	},
	"item": {
		"label": "PowerView Shade",
		"groupNames": [
		]
	},
	"ID": "fdb5d366",
	"label": "PowerView Shade",
	"bridgeUID": "hdpowerview:hub:g24",
	"location": "xxxx"
}{
	"UID": "hdpowerview:shade:fdb5d366",
	"thingTypeUID": "hdpowerview:shade",
	"configuration": {
		"id": "yyyy"
	},
	"item": {
		"label": "PowerView Shade",
		"groupNames": [
		]
	},
	"ID": "fdb5d366",
	"label": "PowerView Shade",
	"bridgeUID": "hdpowerview:hub:g24",
	"location": "xxxx"
}

HTTP/1.1 201 Created
Content-Type: application/json
Content-Encoding: UTF-8
Content-Length: 1412
Server: Jetty(9.4.20.v20190813)

{
	"statusInfo": {
		"status": "OFFLINE",
		"statusDetail": "COMMUNICATION_ERROR"
	},
	"editable": true,
	"label": "PowerView Shade",
	"bridgeUID": "hdpowerview:hub:g24",
	"configuration": {
		"id": "yyyy"
	},
	"properties": {
	},
	"UID": "hdpowerview:shade:fdb5d366",
	"thingTypeUID": "hdpowerview:shade",
	"channels": [
		{
~~~
		}
	],
	"location": "xxxx"
}

```

**Comment by andrewfg** [comment]
^
Looking at the trace, makes me think that the bug is perhaps not in the UI, but rather in the REST API, or perhaps in the core itself?

**Comment by andrewfg** [comment]
![image](https://user-images.githubusercontent.com/893994/87703555-f94dce80-c792-11ea-8b93-811062609550.png)

**Comment by ghys** [comment]
Sure, I remember having this conversation already; this can be fixed, but keep in mind that the parent bridge can be changed after the thing's creation, however the thingUID cannot change... therefore including the bridgeUID in it cannot be considered a reliable reference to the actual parent bridge.

**Comment by andrewfg** [comment]
Hi @ghys many thanks for the response. I get your point that it might break if the bridge would be changed in the future. However my point is that _it is already broken even if the the bridge would **not** be changed_.

The Thing UID naming convention for Things with Bridges is `bridge_type:thing_type:bridge_instance_id:thing_instance_id` with four arguments. But this UI page is creating Things with a UID comprising only three arguments. So in other words, this page is not compliant with the OH naming convention.

The consequence of the missing `bridge_instance_id` is that the framework is forced to create a dysfunctional "zombie" Thing that lacks a reference to a parent bridge instance. The result is that such zombie Things are not (cannot be) initialized, and therefore cannot exchange live channel data with the field.

**Comment by ghys** [comment]
@andrewfg found the conversation I was referring to above:
https://github.com/openhab/openhab-addons/pull/7496#discussion_r426005622
(this comment and the posts above it)

The conclusion that was reached is that it is wrong of add-ons to:
> read any semantics out of a ThingID (because the user might have changed the bridge or the binding migrated the thing type to a different one or whatever might have been happened)

That being said, in order to keep add-ons which do from breaking (until they're hopefully fixed and don't attempt to find a reference to a thing's bridge by parsing its uid), I think it would be acceptable to include the "_first_" bridgeUid i.e. the one that was specified during the thing's creation in the new thing's uid - even if it's not actually mandatory to specify it (it can be done later in the UI via the thing details page) or can be altered while the thingUid has to remain immutable. That's what is done for things added from the Inbox anyway, so at least we'd have a consistent approach on this matter.

**Comment by andrewfg** [comment]
> .. in order to keep add-ons which do from breaking (until they're hopefully fixed and don't attempt to find a reference to a thing's bridge by parsing its uid)

Hmmm. @berland and I have observed this problem with the Hue, HD PowerView, Velux, NeoHub bindings (and since it is based on the NeoHub code, I suspect that my Siemens RDS is also affected). These bindings use the core's `BaseThingHandler.class` method `getBridge()` (see below) to initialize their local reference to their respective bridge / hub. It appears that this method is guilty of the crime of "attempting to find a reference to a thing's bridge by parsing its uid" so I suppose that any binding that calls this inherited method would be equally "guilty" of that crime. => So I suspect that the fix is either in the core or the UI and not in the (many) bindings that depend on the core. Or??

```
    protected @Nullable Bridge getBridge() {
        ThingUID bridgeUID = thing.getBridgeUID();
        synchronized (this) {
            if (callback != null) {
                return bridgeUID != null ? callback.getBridge(bridgeUID) : null;
            } else {
..
            }
        }
    }

```

**Comment by andrewfg** [comment]
> I think it would be acceptable to include the "first" bridgeUid i.e. the one that was specified during the thing's creation in the new thing's uid

@ghys the UI page has a drop down for selecting the bridge instance; so putting aside the discussion in my prior post, I believe that there should be two distinct behaviours depending on whether the user actively makes a selection in this drop down:

- a bridge is selected in the drop down => create the Thing **with** the selected bridge UID in its name
- a bridge is NOT selected in the drop down => create the Thing **without** a bridge UID in its name (current behaviour)

What do you think?

**Comment by ghys** [comment]
Exactly, that was my suggestion and implemented in the proposed fix (#298).
